### DISCUSSION TITLE
Workflow recipe for building app with dependencies

### DISCUSSION THREAD

**Discussion by LucienMorey** [nasa/cFS#563] [Q&A] [answered]
Hi,

I am interested in trying to modify the existing workflow recipes so they accept an additional arg to build additional apps or libs as dependencies. I have managed to do a port of IO_LIB, CI and TO but the latter two depend on each other and on IO_LIB to build and run. Does anyone know how this might be done or have an existing example I could use?

I think it might be possible to create a json from an input string and then turn that into a matrix for use. I am totally new to writing my own workflows and haven't wrapped my head around the documentation on how this might be done yet. It would also be helpful if anyone has some ideas on how to do this or if it's even a reasonable proposal.

**Comment by skliper** [comment, marked as answer]
The only reusable workflow that would be necessary to actually check out dependencies is the build and run, right?  I think the only three changes would be to check out the additional repo, add to startup, add to build.  I'm not sure how to easily handle a matrix of dependencies... although if you just wanted to handle one you could simply add conditional steps based on an additional input.

**Reply by LucienMorey** [reply]
I would think unit test coverage requires this too to link stubs properly. Right now I link them in the unit test CMake where the CFE stubs are linked. Is this incorrect?

No worries about the matrix thing. I might look into it more if I get some free time. I am currently adding the additional dependencies manually to a modified copy of the workflow, but I figured I would try to reuse as much as possible without doubling up. I will leave this open in case someone else has a brainwave that can solve the problem more elegantly.

**Reply by skliper** [reply]
Good point, definitely need the stubs for ut.  Maybe consider adding an action for each dependency (implemented by the dependency) that could add what's needed, to minimize the complexity additions to the reusable workflow?  I haven't messed with actions much, but seems like a good way to abstract the details.  I haven't found anything for running a step over a matrix or loop a step, although there's questions out there: https://stackoverflow.com/questions/59180385/using-an-array-of-values-to-repeat-a-step-in-github-actions-workflow.

**Reply by LucienMorey** [reply]
I checked into this more but I don't think matrix operations are the right way to go about this anymore.

The solution I went with was to:
- Add all my apps as submodules so they could each be cloned in a runner
- Override the app that triggered the workflow with the branch that was pushed to
- Add basic ```targets.cmake``` and ```cpu1_cfe_es_startup.scr``` to control the build so an app and its dependencies are built exclusively
- add an additional argument to the reusable workflows to specify if the intended target is an app or lib

The rest of the workflow remains more or less the same as what exists currently. I figure there should be no issues with circular dependencies when building targets if things are split up properly.

Hopefully, this is a reasonable solution for anyone else interested, but I would love to hear of a better one if anyone works on this in the future.
